### Q1. Pandas version

What's the version of Pandas that you installed?

You can get the version information using the `__version__` field:

In [24]:
import numpy as np
import pandas as pd

In [25]:
pd.__version__

'2.3.2'

In [26]:
df = pd.read_csv("data/car_fuel_efficiency.csv")
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

### Q2. Records count

How many records are in the dataset?

- 4704
- 8704
- 9704
- 17704

In [27]:
print(f"The total records in the dataset is: {len(df)}")

The total records in the dataset is: 9704


### Q3. Fuel types

How many fuel types are presented in the dataset?

- 1
- 2
- 3
- 4

In [28]:
unique_fuel_types = df["fuel_type"].nunique()
print(f"The total unique fuel types are: {unique_fuel_types}")

The total unique fuel types are: 2


### Q4. Missing values
How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3
- 4

In [29]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [30]:
rows_w_null = (df.isnull().sum() > 0).sum()
print(f"There are {rows_w_null} rows with missing values.")

There are 4 rows with missing values.


### Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

- 13.75
- 23.75
- 33.75
- 43.75

In [31]:
max_fuel_efficiency_asian_cars = df[df["origin"] == "Asia"]["fuel_efficiency_mpg"].max()
print(f"The max fuel efficiency for asian cars is {max_fuel_efficiency_asian_cars:.2f}")

The max fuel efficiency for asian cars is 23.76


### Q6. Median value of horsepower
1. Find the median value of horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.

Has it changed?

- Yes, it increased
- Yes, it decreased
- No

In [32]:
# 1. Find the median value of horsepower column in the dataset.
median_horsepower = df["horsepower"].median()
print(f"The median value of horsepower is {median_horsepower:.2f}")

# 2. Next, calculate the most frequent value of the same horsepower column.
most_frequent_horsepower = df["horsepower"].mode()[0]
print(f"The most frequent value of horsepower is {most_frequent_horsepower:.2f}")

# 3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
df["horsepower"] = df["horsepower"].fillna(most_frequent_horsepower)

median_horsepower = df["horsepower"].median()
print(f"The median value of horsepower is {median_horsepower:.2f}")

The median value of horsepower is 149.00
The most frequent value of horsepower is 152.00
The median value of horsepower is 152.00


### Q7. Sum of weights

1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.051
- 0.51
- 5.1
- 51

In [33]:
# 1. Select all the cars from Asia.
asia_cars = df[df["origin"] == "Asia"]
asia_cars

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0       152.0     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [34]:
# 2. Select only columns `vehicle_weight` and `model_year`2. Select only columns `vehicle_weight` and `model_year`
weight_n_year = asia_cars[["vehicle_weight", "model_year"]]
weight_n_year

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [35]:
# 3. Select the first 7 values
first_7_values = weight_n_year.head(7)
first_7_values

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [36]:
# 4. Get the underlying NumPy array. Let's call it `X`.
X = first_7_values.to_numpy()
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [37]:
# 5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.

XTX = X.T @ X
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [38]:
# 6. Invert `XTX`.
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [39]:
# 7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [40]:
# 8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
w = XTX_inv.dot(X.T.dot(y))
# w = XTX_inv @ X.T @ y
print("w =", w)

w = [0.01386421 0.5049067 ]


In [41]:
# 9. What's the sum of all the elements of the result?
sum_w = w.sum()        # or np.sum(w)
print("Sum of elements in w =", sum_w)

Sum of elements in w = 0.5187709081074022
